packages/react/dismissable-layer/src/DismissableLayer.tsx

```diff
@@ -1,241 +1,167 @@
-    // Only dismiss if there's no deeper layer which disabled pointer events outside itself
-    if (!containsChildLayerWithDisabledOutsidePointerEvents) {
+const DismissableLayer = React.forwardRef<DismissableLayerElement, DismissableLayerProps>(
+  (props, forwardedRef) => {
+    const {
+      disableOutsidePointerEvents = false,
+      onEscapeKeyDown,
+      onPointerDownOutside,
+      onFocusOutside,
+      onInteractOutside,
+      onDismiss,
+      ...layerProps
+    } = props;
+    const context = React.useContext(DismissableLayerContext);
+    const [node, setNode] = React.useState<DismissableLayerElement | null>(null);
+    const [, force] = React.useState({});
+    const composedRefs = useComposedRefs(forwardedRef, (node) => setNode(node));
+    const [highestModal] = [...context.modals].slice(-1);
+    const highestModalIndex = [...context.layers].indexOf(highestModal);
+    const layerIndex = node ? [...context.layers].indexOf(node) : -1;
+    const hasPointerEvents = layerIndex >= highestModalIndex;
+
+    const pointerDownOutside = usePointerDownOutside((event) => {
+      if (!hasPointerEvents) return;
       onPointerDownOutside?.(event);
       onInteractOutside?.(event);
-      if (!event.defaultPrevented) {
-        onDismiss?.();
-      }
-    }
-  });
+      if (!event.defaultPrevented) onDismiss?.();
+    });
 
-  // Dismiss on focus outside
-  const { onBlurCapture: handleBlurCapture, onFocusCapture: handleFocusCapture } = useFocusOutside(
-    (event) => {
+    const focusOutside = useFocusOutside((event) => {
       onFocusOutside?.(event);
       onInteractOutside?.(event);
-      if (!event.defaultPrevented) {
-        onDismiss?.();
-      }
-    }
-  );
+      if (!event.defaultPrevented) onDismiss?.();
+    });
 
-  // If we have disabled pointer events on body, we need to reset `pointerEvents: 'auto'`
-  // on some layers. This depends on which layers set `disableOutsidePointerEvents` to `true`.
-  //
-  // NOTE: it's important we set it on ALL layers that need it as we cannot simply
-  // set it on the deepest layer which sets `disableOutsidePointerEvents` to `true` and rely
-  // on inheritence. This is because layers may be rendered in different portals where
-  // inheritence wouldn't apply, so we need to set it explicity on its children too.
-  const isBodyPointerEventsDisabled = totalLayerCountWithDisabledOutsidePointerEvents > 0;
+    useEscapeKeydown((event) => {
+      const isHighestLayer = layerIndex === context.layers.size - 1;
+      if (!isHighestLayer) return;
+      onEscapeKeyDown?.(event);
+      if (!event.defaultPrevented) onDismiss?.();
+    });
+
+    useBodyPointerEvents({ disabled: disableOutsidePointerEvents });
+
+    React.useEffect(() => {
+      if (!node) return;
+      if (disableOutsidePointerEvents) context.modals.add(node);
+      context.layers.add(node);
+      dispatchUpdate();
+    }, [node, disableOutsidePointerEvents, context]);
+
+    React.useEffect(() => {
```

Comment:
What is preventing these 2 effects be combined? as the second should just be the cleanup right, ie. when a layer is removed, or `disableOutsidePointerEvents` is changed.

Reply:
A change to `disableOutsidePointerEvents` would remove from the stack and add it to the end again meaning the layering order wouldn't be "creation order" anymore. I only want them to be removed from layer/modal stacks when unmounted.

Reply:
Ah that's because `Set` guarantees insertion order?

Reply:
it's just the same as an array right, but prevents duplicates

Reply:
Yeah I always forget this somehow!